File: src/services/mywod.ts
import * as fs from 'fs';
import * as path from 'path';
import * as mongoose from 'mongoose';
import * as sqlite from 'sqlite';
import * as HttpStatus from 'http-status-codes';
import * as _ from 'lodash';

import { UserModel, UserType } from '../models/user';
import { WorkoutModel, WorkoutType } from '../models/workout';
import { MovementModel, MovementType } from '../models/movement';
import { MovementScoreModel, MovementScoreType } from '../models/movement.score';
import { WorkoutScoreModel, WorkoutScoreType } from '../models/workout.score';
import ExpressError from '../utils/express.error';
import { MywodUtils } from '../utils/mywod.utils';

export class MywodService {
	public static FILE_LOCATION = `${process.cwd()}/mywod`;
	private userModel: mongoose.Model<UserType>;
	private workoutModel: mongoose.Model<WorkoutType>;
	private workoutScoreModel: mongoose.Model<WorkoutScoreType>;
	private movementModel: mongoose.Model<MovementType>;
	private movementScoreModel: mongoose.Model<MovementScoreType>;

	constructor(public options: any = {}) {
		this.userModel = this.options.userModel || new UserModel().createModel();
		this.workoutModel = this.options.workoutModel || new WorkoutModel().createModel();
		this.workoutScoreModel = this.options.workoutScoreModel || new WorkoutScoreModel().createModel();
		this.movementModel = this.options.movementModel || new MovementModel().createModel();
		this.movementScoreModel = this.options.movementScoreModel || new MovementScoreModel().createModel();
	}

	public async saveAthlete(user: any, data: any) {
		if (data.email !== user.email) {
			throw new ExpressError('Emails do not match', 'This myWOD backup does not belong to this email address', HttpStatus.FORBIDDEN);
		}

		let model = await this.userModel.findOne({ email: user.email });

		if (!model) {
			throw new ExpressError('Not found', 'Could not detect a user logged in', HttpStatus.NOT_FOUND);
		}

		model = this.updateUser(model, data);
		return model.save();
	}

	private updateUser(model: UserType, data: any) {
		model.firstName = data.firstName;
		model.lastName = data.lastName;
		model.height = data.height;
		model.weight = data.weight;
		model.gender = data.gender;
		model.dateOfBirth = data.dateOfBirth;
		model.boxName = data.boxName;
		return model;
	}

	public async saveWorkouts(user: UserType, workouts: any[]) {
		const savedWorkouts: any[] = [];
		for (const workout of workouts) {
			try {
				if (workout.description.startsWith('This is a sample custom WOD')) {
					continue; // Ignore the sample wod made by myWOD
				}
				workout.createdBy = user._id;
				workout.measurement = workout.scoreType;
				const workoutModelInstance = new this.workoutModel(workout);
				await workoutModelInstance.save();
				savedWorkouts.push(workoutModelInstance.title);
			} catch (err) {
				console.log(`Error migrating workout ${workout.title}. Error: ${err}`);
			}
		}

		return savedWorkouts;
	}

	public async saveWorkoutScores(user: UserType, workoutScores: any[]) {
		const scoresSorted = _.sortBy(workoutScores, ['title']);
		for (const score of scoresSorted) {
			try {
				const scoreData = MywodUtils.parseWorkoutScore(score);
				const scoreModelInstance = new this.workoutScoreModel(scoreData);
				scoreModelInstance.createdBy = user._id;
				await scoreModelInstance.save();

				const workoutModel = await this.workoutModel.findOne({ 'title': score.title, 'createdBy': user._id });
				if (workoutModel) {
					workoutModel.scores.push(scoreModelInstance._id);
					await workoutModel.save();
				}
			} catch (err) {
				console.log(`Could not migrate score for ${score.title}`);
			}
		}
	}

	public async saveMovementsAndMovementScores(user: UserType, movements: any[], movementScores: any[]) {
		const savedMovements: string[] = [];

		for (const movement of movements) {
			try {
				movement.createdBy = user._id;
				movement.measurement = movement.type;
				const movementModelInstance = new this.movementModel(movement);

				const scores = MywodUtils.getScoresForMovement(movement, movementScores);
				await this.saveScoresForMovement(user, movement, movementModelInstance, scores);
				savedMovements.push(movementModelInstance.name);
			} catch (err) {
				console.log(`Error migrating movement '${movement.name}'. Error: ${err}`);
			}
		}
		return savedMovements;
	}

	private async saveScoresForMovement(user: UserType, movement: any, movementModelInstance: MovementType, scores: any[]) {
		for (const score of scores) {
			try {
				const scoreModelInstance = new this.movementScoreModel(score);
				scoreModelInstance.createdBy = user._id;
				await scoreModelInstance.save();
				movementModelInstance.scores.push(scoreModelInstance._id);
			} catch (err) {
				console.error(`Error migrating movement score '${score.score}' for '${movement.name}'`);
			}
		}
		return movementModelInstance.save();
	}

	public async readContentsFromDatabase(filename: string) {
		const resolvedPath = this.resolvePath(filename);
		const db = await sqlite.open(resolvedPath);

		const athlete = await db.get('SELECT * FROM Athlete LIMIT 1;');
		const workouts = await db.all('SELECT * FROM CustomWODs;');
		const movements = await db.all('SELECT * FROM Movement;');
		const movementScores = await db.all('SELECT * FROM MovementSessions;');
		const workoutScores = await db.all('SELECT * FROM MyWODs;');

		return {
			athlete,
			workouts,
			movements,
			movementScores,
			workoutScores
		};
	}

	public deleteDatabaseFile(filename: string) {
		const resolvedPath = this.resolvePath(filename);
		fs.unlinkSync(resolvedPath);
	}

	public resolvePath(filename: string) {
		return path.resolve(process.cwd(), MywodService.FILE_LOCATION, filename);
	}
}

File: src/services/mywod.rs
use crate::errors::WebResult;
use crate::models::movement::CreateMovement;
use crate::models::mywod::{Athlete, CustomWOD, Movement, MovementSession, MyWOD};
use crate::models::user::UpdateUser;
use crate::models::workout::CreateWorkout;
use crate::repositories::{MovementRepository, UserRepository, WorkoutRepository};
use crate::utils::mywod::{
    get_scores_for_movement, map_movement_measurement, map_workout_measurement,
    parse_workout_score, save_avatar,
};

pub async fn save_athlete(
    user_repo: UserRepository,
    user_id: &str,
    user_email: &str,
    athlete: Athlete,
) -> WebResult<bool> {
    let avatar_url = save_avatar(user_id, athlete.avatar).unwrap();

    let user_update = UpdateUser {
        password: None,
        first_name: Some(athlete.first_name.trim().to_owned()),
        last_name: Some(athlete.last_name.trim().to_owned()),
        date_of_birth: Some(athlete.date_of_birth),
        height: Some(athlete.height),
        weight: Some(athlete.weight),
        box_name: Some(athlete.box_name.trim().to_owned()),
        avatar_url: Some(avatar_url),
    };

    let _ = user_repo
        .update_user_with_email(user_email, user_update)
        .await?;

    Ok(true)
}

pub async fn save_workouts_and_scores(
    workout_repo: WorkoutRepository,
    workouts: Vec<CustomWOD>,
    workout_scores: &[MyWOD],
    user_id: &str,
) -> WebResult<(u32, u32)> {
    let mut added_workouts = 0u32;
    let mut added_scores = 0u32;

    for workout in workouts {
        debug!("Processing workout '{}'", workout.title);
        let new_workout = CreateWorkout {
            name: workout.title.to_owned(),
            description: workout.description,
            measurement: map_workout_measurement(workout.score_type.as_ref()),
            is_public: false,
        };
        let created_workout = workout_repo.create_workout(user_id, new_workout).await;

        match created_workout {
            Ok(_) => added_workouts += 1,
            Err(e) => warn!(
                "Could not create new workout from backup \"{}\". Error: {}",
                workout.title,
                e.to_string()
            ),
        }
    }

    // TODO: Fix duplication with workout scores
    for score in workout_scores {
        let workout_title = score.title.to_owned();
        let workout_description = score.description.to_owned();

        let mut workout = workout_repo
            .find_workout_by_name(user_id, &workout_title)
            .await?;

        if workout.is_none() {
            let new_workout = CreateWorkout {
                name: workout_title,
                description: workout_description,
                measurement: map_workout_measurement(&score.score_type),
                is_public: false,
            };
            workout = Some(workout_repo.create_workout(user_id, new_workout).await?);
            added_workouts += 1;
        }

        if let Some(workout) = workout {
            debug!("Processing scores for '{}", workout.name);
            let score_data = parse_workout_score(score);
            let added_score = workout_repo
                .create_workout_score(user_id, &workout, score_data)
                .await;
            if added_score.is_ok() {
                added_scores += 1;
            }
        }
    }

    Ok((added_workouts, added_scores))
}

pub async fn save_movements_and_scores(
    movement_repo: MovementRepository,
    movements: &[Movement],
    movement_scores: &[MovementSession],
    user_id: &str,
) -> WebResult<(u32, u32)> {
    let mut added_movements = 0u32;
    let mut added_movement_scores = 0u32;

    for m in movements {
        let new_movement = CreateMovement {
            name: m.name.to_owned(),
            measurement: map_movement_measurement(m.score_type),
            is_public: false,
        };
        let created_movement = movement_repo.create_movement(user_id, new_movement).await;
        match created_movement {
            Ok(created_movement) => {
                added_movements += 1;
                let all_scores = get_scores_for_movement(m, movement_scores);
                for score in all_scores {
                    movement_repo
                        .create_movement_score(user_id, &created_movement, score)
                        .await?;
                    added_movement_scores += 1;
                }
            }
            Err(e) => warn!(
                "Could not add movement {}. Error: {}",
                m.name,
                e.to_string()
            ),
        }
    }

    Ok((added_movements, added_movement_scores))
}
